### PR TITLE
populate headers to client request

### PR DESCRIPTION
**Pull request by cl1337**
this PR add build time middleware to assign headers to client request args

* for any client arg, it should either come from endpoint request / endpoint request transform, or headers with header transform config
* support override from headers 
* currently only support header values as string, no type convert now (like `version: 3` -> `int64(3)`)

**example** (also check `trans-headers.json` in example-gateway)

endpoint thrift
```
  struct TransHeader {}

  TransHeader transHeaders(
      1: required TransHeader req
  ) throws (
      1: AuthErr authErr (zanzibar.http.status = "401")
      2: OtherAuthErr otherAuthErr (zanzibar.http.status = "403")
  ) (
      zanzibar.http.status = "200"
      zanzibar.http.method = "POST"
      zanzibar.http.path = "/baz/trans-headers"
      zanzibar.http.reqHeaders = "x-uuid,x-token"
  )
```

client thrift
```
struct NestHeaders {
    1: required string uuid
    2: optional string token
}

struct Wrapped {
    1: required NestHeaders n1
    2: optional NestHeaders n2
}

struct TransHeaders {
    1: required Wrapped w1
    2: optional Wrapped w2
}
```
```
base.TransHeaders transHeaders(
    1: required base.TransHeaders req
) throws (
    1: AuthErr authErr
    2: OtherAuthErr otherAuthErr
)
```

set header populator config
```javascript
"middlewares": [
   {
       "name": "headersPropagate",
       "options": {
           "propagate": [
               {
                   "from": "x-uuid",
                   "to": "Req.W1.N1.UUID",
                   "override": true
               },
               {
                   "from": "x-uuid",
                   "to": "Req.W1.N2.UUID",
                   "override": false
              },
               {
                  "from": "x-uuid",
                   "to": "Req.W2.N1.UUID",
                   "override": false
               },
               {
                   "from": "x-uuid",
                   "to": "Req.W2.N2.UUID",
                   "override": true
               },
               {
                   "from": "x-token",
                   "to": "Req.W1.N1.Token",
                   "override": true
               },
               {
                   "from": "x-token",
                   "to": "Req.W1.N2.Token",
                   "override": false
               },
               {
                   "from": "x-token",
                   "to": "Req.W2.N1.Token",
                   "override": false
               },
               {
                   "from": "x-token",
                   "to": "Req.W2.N2.Token",
                   "override": true
               }
           ]
       }
   }
]
```

### CODE REVIEW

File: codegen/gateway.go [hunk 1]
```diff
@@ -777,6 +789,44 @@ func augmentHTTPEndpointSpec(
...
+				"populate middleware found with no source field.",
+			)
+		}
+		toField, ok := populateMap["to"].(string)
+		if !ok {
+			return nil, errors.Errorf(
+				"populate middleware found with no destination field.",
+			)
+		}
+		overrideOpt, ok := populateMap["override"].(bool)
```
Comment (ChuntaoLu) [inline codegen/gateway.go]: It should error explicitly  if user didn't set a boolean value for the `override` field.

File: codegen/gateway.go [hunk 2]
```diff
@@ -777,6 +789,44 @@ func augmentHTTPEndpointSpec(
...
+		if ok {
```
Comment (ChuntaoLu) [inline codegen/gateway.go]: This if else block is unnecessary.
```go
fieldMap[toField] = FieldMapperEntry{
        QualifiedName: fromField,
        Override: overrideOpt,
}
```
is sufficient.

File: codegen/header_populator.go [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+			}
+		}
+		if len(prevPath) == len(currPath) {
+			return nil, missErr
+		}
+	}
+	return nil, missErr
+}
+
+func (hp *HeaderPopulator) getIdentifierName(fieldType compile.TypeSpec) (string, error) {
```
Comment (ChuntaoLu) [inline codegen/header_populator.go]: This function is simple and only has one call site, doesn't seem necessary.

File: examples/example-gateway/idl/endpoints/baz/baz.thrift [hunk 4]
```diff
@@ -63,6 +65,18 @@ service SimpleService {
       zanzibar.handler = "baz.trans"
   )
 
+  TransHeader transHeaders(
+      1: required TransHeader req
```
[the remark this answers is not in the document]
Comment (cl1337) [reply examples/example-gateway/idl/endpoints/baz/baz.thrift]: no, `TransHeader ` is empty struct that does not contain params required for clients

endpoint params come from `zanzibar.reqHeaders` annotation, you don't have to change existing endpoint request params to populate, but annotation is required

**Comment by cl1337** [comment]
^ @oliviazhang0809  checkout test example in `examples/example-gateway/build/gen-code/clients/baz/baz/simpleservice_transheaders.go`

File: codegen/type_converter.go [hunk 5]
```diff
@@ -73,15 +73,17 @@ type TypeConverter struct {
 	fieldCounter  int
 	convStructMap map[string]string
 	useRecurGen   bool
+	auxChecks     map[string]FieldMapperEntry
 }
 
 // NewTypeConverter returns *TypeConverter
-func NewTypeConverter(h PackageNameResolver) *TypeConverter {
+func NewTypeConverter(h PackageNameResolver, auxChecks map[string]FieldMapperEntry) *TypeConverter {
```
Comment (ChuntaoLu) [inline codegen/type_converter.go]: `optionalEntries` feels like more indicative to what `auxChecks` does, also let's put a comment for the function argument and explain the intent.

File: codegen/gateway.go (hunk 2 above)
Comment (cl1337) [reply codegen/gateway.go]: think your above 2 comments on same block:

I'll go with raising error, `override` has to be configured as bool, no fallback / error swallow -> no undefined behavior

File: codegen/gateway.go [hunk 6]
```diff
@@ -429,6 +429,9 @@ type EndpointSpec struct {
 	TestFixtures map[string]*EndpointTestFixture
 	// Middlewares, meta data to add middlewares,
 	Middlewares []MiddlewareSpec
+	// HeadersIntoReq, a map from endpoint request headers to
+	// client request fields.
+	HeadersIntoReq map[string]FieldMapperEntry
```
Comment (ChuntaoLu) [inline codegen/gateway.go]: If we change this to `HeadersPropagate`, then we can s/populate/propagate everywhere and be consistent.

File: examples/example-gateway/middlewares/headers-propagate/headers_propagate_schema.json [hunk 7]
```diff
@@ -0,0 +1,23 @@
+{
```
Comment (cl1337) [inline examples/example-gateway/middlewares/headers-propagate/headers_propagate_schema.json]: added schema template

File: examples/example-gateway/idl/clients/baz/base.thrift [hunk 8]
```diff
@@ -17,6 +17,21 @@ struct TransStruct {
     3: required NestedStruct rider
 }
 
+struct NestHeaders {
+    1: required string UUID
```
Comment (cl1337) [inline examples/example-gateway/idl/clients/baz/base.thrift]: add typedef string case

File: codegen/header_propagate_test.go [hunk 9]
```diff
@@ -139,6 +139,47 @@ func TestDefault(t *testing.T) {
 	assert.Equal(t, strip(s), strip(lines))
 }
 
+func TestTypeDefString(t *testing.T) {
```
Comment (cl1337) [inline codegen/header_propagate_test.go]: add a new test case for `typedef string UUID`

File: codegen/gateway.go [hunk 10]
```diff
@@ -796,6 +808,44 @@ func augmentHTTPEndpointSpec(
...
+func setPropagateMiddleware(middlewareObj map[string]interface{}) (map[string]FieldMapperEntry, error) {
+	fieldMap := make(map[string]FieldMapperEntry)
+	opts, ok := middlewareObj["options"].(map[string]interface{})
+	if !ok {
+		return nil, errors.New(
+			"missing or invalid options for propagate middleware",
+		)
+	}
+	propagates := opts["propagate"].([]interface{})
+	dest := make(map[string]string)
```
Comment (cl1337) [inline codegen/gateway.go]: adding checks to prevent multiple assignment to same destination

File: codegen/header_propagate.go [hunk 11]
```diff
@@ -0,0 +1,151 @@
...
+		// future support:
+		// e.g: -H "version: 3.0" -> int64(3)
+
+		assignedVal := "key"
+		if gotype != "string" {
+			assignedVal = gotype + "(key)"
+		}
+
+		if !field.Required {
+			hp.appendf(`in.%s = &%s`, key, assignedVal)
```
Comment (cl1337) [inline codegen/header_propagate.go]: remove override as an option

File: codegen/gateway_test.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+package codegen
+
+import (
+	"encoding/json"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestHeadersPropagateMultiDest(t *testing.T) {
```
Comment (cl1337) [inline codegen/gateway_test.go]: add test for multiple destination assignment

**Pull request merged**